core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -364,9 +365,14 @@ class TransactionCoordinator(brokerId: Int,
           val coordinatorEpoch = epochAndTxnMetadata.coordinatorEpoch
 
           txnMetadata.inLock {
-            if (txnMetadata.producerId != producerId)
+            // If we are ending the transaction due to a fenced producer and we are rotating the producer ID because the
+            // epoch was exhausted, we need to skip the typical ID and epoch checks because the transaction metadata
+            // object will already have been changed.
+            val isFencedProducerRotation = txnMetadata.isProducerEpochExhausted &&
```

Comment:
Do all of these checks make sense for an EndTxnRequest from the client? I guess we could get a call to EndTxn from a zombie producer while it is in the middle of being fenced. I think the checks below would result in `CONCURRENT_TRANSACTIONS` which would cause the zombie to retry until the transition completes. Maybe that's reasonable?